openpype/settings/defaults/project_settings/global.json

```diff
@@ -397,11 +397,18 @@
             ]
         },
         "CleanUp": {
```

Comment:
Why are these optional? Optional means artist (user) can change it's 'active' state before publishing. Some of these are headless plugins -> there is no room to change it's state. And also I'm not sure if it's good idea to give user the power to affect that?

Reply:
> Yeah good call, is optional even working in the new publisher though?

Yes they work, but in a different way.

> and made it consistent with the rest of clean up plugins

I don't know why this one is optional, but it does not make sence with the other plugins. Farm cleanup cannot be changed -> doesn't make sence to have it optional, and explicit cleanup is for files that are 100% to remove (they are in temp or cannot be reused).

Reply:
> Yes they work, but in a different way.

What different way?

> explicit cleanup is for files that are 100% to remove (they are in temp or cannot be reused) -> even advanced users would not want that.

In Nuke farm renders it's removing the files from the temporary staging directory, we don't want those removed so that's why we want to have control of disabling the plugin

I'm going to go ahead and remove the optionals

Reply:
Done here https://github.com/ynput/OpenPype/pull/5141/commits/08671cb49d830a7cdc0ca1413b1db59cde3ecef3

Reply:
> What different way?

Already described, one is on farm -> No way how to change state. Explicit Cleanup plugin is designed to explicitly remove files by pyblish plugins.

> In Nuke farm renders it's removing the files from the temporary staging directory, we don't want those removed so that's why we want to have control of disabling the plugin

That's a use-case that deservers changes, BUT that should be handled in Nuke plugin which is causing that, not in these plugins. These are just not meant to be optional. Single usecase should not affect every other. "Keep rendered files once done" should not cause that temp is flooded by other temporary files.

Reply:
> Already described, one is on farm -> No way how to change state -> no reason to have setting for that. Explicit Cleanup plugin is designed to explicitly remove files by pyblish plugins.

I meant how is the `Optional` working differently on the plugins on the new publisher, you are not responding to that. How is the user able to enable/disable the optional plugins on the newp ublisher?

> That's a use-case that deservers changes, BUT that should be handled in Nuke plugin which is causing that, not in these plugins. These are just not meant to be optional. Single usecase should not affect every other. "Keep rendered files once done" should not cause that temp is flooded by other temporary files from ExtractReview, ExtractBurnin and others.

Fair enough, I can investigate where are these files being explicitly added to be deleted but I think there's still benefit on exposing control on the plugins you want to execute or not, specially the ones that delete data

Reply:
> I meant how is the Optional working differently on the plugins on the new publisher, you are not responding to that. How is the user able to enable/disable the optional plugins on the new publisher?

The plugin must inherit from `OptionalPyblishPluginMixin` and have to have this at the start of process.
```
def process(self, instance):
    if not self.is_active(instance.data):
        return
```

Reply:
> > I meant how is the Optional working differently on the plugins on the new publisher, you are not responding to that. How is the user able to enable/disable the optional plugins on the new publisher?
> 
> The plugin must inherit from `OptionalPyblishPluginMixin` and have to have this at the start of process.
> 
> ```python
> def process(self, instance):
>     if not self.is_active(instance.data):
>         return
> ```

I will repeat, how is the **user** able to enable/disable the optional plugins on the new publisher? 

That would need to be through the UI

Reply:
He can enabled/disable it per instance (if it's instance plugin) on Publish tab in right panel. If it's context plugin then he can enable/disable it when Context is selected.